File: server/app/utils/sentry.py
"""Sentry integration for error tracking and monitoring."""

import os
import sentry_sdk
from sentry_sdk.integrations.flask import FlaskIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.redis import RedisIntegration
from sentry_sdk.integrations.logging import LoggingIntegration
from flask import request


def init_sentry(app):
    """Initialize Sentry for the Flask application."""
    dsn = os.getenv('SENTRY_DSN')
    
    if not dsn:
        app.logger.info('Sentry DSN not configured, skipping initialization')
        return
    
    # Configure logging integration
    logging_integration = LoggingIntegration(
        level=None,  # Capture all levels
        event_level=None  # Don't create events from logs by default
    )
    
    sentry_sdk.init(
        dsn=dsn,
        integrations=[
            FlaskIntegration(
                transaction_style='endpoint'
            ),
            SqlalchemyIntegration(),
            RedisIntegration(),
            logging_integration,
        ],
        
        # Performance monitoring
        traces_sample_rate=float(os.getenv('SENTRY_TRACES_SAMPLE_RATE', 0.1)),
        
        # Environment
        environment=app.config.get('ENV', 'production'),
        
        # Release tracking
        release=os.getenv('APP_VERSION', '1.0.0'),
        
        # Configure error filtering
        before_send=before_send_filter,
        
        # Attach request data
        attach_stacktrace=True,
        send_default_pii=False,  # Don't send personally identifiable information
        
        # Performance
        profiles_sample_rate=float(os.getenv('SENTRY_PROFILES_SAMPLE_RATE', 0.1)),
    )
    
    app.logger.info('Sentry initialized successfully')


def before_send_filter(event, hint):
    """Filter events before sending to Sentry."""
    # Don't send events in development unless explicitly enabled
    if os.getenv('FLASK_ENV') == 'development' and not os.getenv('SENTRY_ENABLED'):
        return None
    
    # Filter out health check requests
    if 'request' in event and event['request'].get('url', '').endswith('/health'):
        return None
    
    # Filter out 404 errors for common bot paths
    if 'exception' in event:
        for value in event['exception']['values']:
            if value.get('type') == 'NotFound':
                url = event.get('request', {}).get('url', '')
                bot_paths = ['.php', '.asp', 'wp-admin', 'wp-login', '.env']
                if any(path in url for path in bot_paths):
                    return None
    
    # Add custom context
    if request:
        event['extra']['request_id'] = request.headers.get('X-Request-ID')
        event['extra']['user_agent'] = request.headers.get('User-Agent')
    
    return event


def set_user_context(user):
    """Set user context for Sentry."""
    if not user:
        sentry_sdk.set_user(None)
        return
    
    sentry_sdk.set_user({
        'id': user.id,
        'email': user.email,
        'username': user.username,
        'role': user.role,
    })
    
    # Add additional context
    sentry_sdk.set_context('user_details', {
        'tenant_id': user.tenant_id,
        'is_active': user.is_active,
        'created_at': user.created_at.isoformat() if user.created_at else None,
    })


def capture_message(message, level='info', **kwargs):
    """Capture a message in Sentry."""
    sentry_sdk.capture_message(message, level=level, **kwargs)


def capture_exception(error, **kwargs):
    """Capture an exception in Sentry."""
    sentry_sdk.capture_exception(error, **kwargs)


def add_breadcrumb(message, category='action', level='info', data=None):
    """Add a breadcrumb for better error context."""
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data or {}
    )


def start_transaction(name, op='http.server'):
    """Start a performance transaction."""
    return sentry_sdk.start_transaction(name=name, op=op)


def configure_scope_middleware(app):
    """Configure Sentry scope for each request."""
    @app.before_request
    def configure_sentry_scope():
        """Configure Sentry scope with request information."""
        with sentry_sdk.configure_scope() as scope:
            # Add request information
            scope.set_tag('endpoint', request.endpoint)
            scope.set_tag('method', request.method)
            
            # Add custom headers
            scope.set_extra('request_id', request.headers.get('X-Request-ID'))
            scope.set_extra('api_version', request.headers.get('X-API-Version', 'v1'))
            
            # Add user context if authenticated
            if hasattr(request, 'current_user') and request.current_user:
                set_user_context(request.current_user)

File: client/src/lib/sentry.js
// TODO: i18n - processed
import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';
/**
 * Initialize Sentry for error tracking and performance monitoring
 */import { useTranslation } from "react-i18next";
export function initSentry() {
  const environment = import.meta.env.MODE;
  const dsn = import.meta.env.VITE_SENTRY_DSN;
  if (!dsn) {
    return;
  }
  Sentry.init({
    dsn,
    environment,
    integrations: [
    new BrowserTracing({
      // Set tracing origins to your frontend and backend
      tracingOrigins: [
      'localhost',
      /^https:\/\/.*\.bdc\.com/,
      /^http:\/\/localhost:\d+/],

      // Performance Monitoring
      routingInstrumentation: Sentry.reactRouterV6Instrumentation(
        React.useEffect,
        useLocation,
        useNavigationType,
        createRoutesFromChildren,
        matchRoutes
      )
    })],

    // Performance Monitoring
    tracesSampleRate: environment === 'production' ? 0.1 : 1.0,
    // Session Replay
    replaysSessionSampleRate: environment === 'production' ? 0.1 : 0,
    replaysOnErrorSampleRate: 1.0,
    // Release tracking
    release: import.meta.env.VITE_APP_VERSION || '1.0.0',
    // User context
    beforeSend(event, hint) {
      // Don't send events in development unless explicitly enabled
      if (environment === 'development' && !import.meta.env.VITE_SENTRY_ENABLED) {
        return null;
      }
      // Filter out some common non-critical errors
      if (event.exception) {
        const error = hint.originalException;
        // Filter out network errors that are expected
        if (error?.message?.includes('Network request failed') &&
        error?.message?.includes('/api/health')) {
          return null;
        }
        // Filter out canceled requests
        if (error?.name === 'AbortError') {
          return null;
        }
      }
      return event;
    },
    // Add user context
    initialScope: {
      tags: {
        component: 'frontend'
      }
    }
  });
}
/**
 * Set user context for Sentry
 */
export function setSentryUser(user) {
  if (!user) {
    Sentry.setUser(null);
    return;
  }
  Sentry.setUser({
    id: user.id,
    email: user.email,
    username: user.username,
    role: user.role
  });
}
/**
 * Log a message to Sentry
 */
export function logToSentry(message, level = 'info', extra = {}) {
  Sentry.captureMessage(message, {
    level,
    extra,
    tags: {
      manual: true
    }
  });
}
/**
 * Capture an exception in Sentry
 */
export function captureException(error, context = {}) {
  console.error('Error captured:', error);
  Sentry.captureException(error, {
    contexts: {
      app: context
    }
  });
}
/**
 * Create a Sentry transaction for performance monitoring
 */
export function startTransaction(name, op = 'navigation') {
  return Sentry.startTransaction({ name, op });
}
/**
 * Add breadcrumb for better error context
 */
export function addBreadcrumb(message, category = 'action', data = {}) {
  Sentry.addBreadcrumb({
    message,
    category,
    level: 'info',
    data,
    timestamp: Date.now() / 1000
  });
}